### DISCUSSION TITLE
Do activex dlls support standard modules or just classes?

### DISCUSSION THREAD

**Discussion by Greedquest** [twinbasic/twinbasic#1456] [Q&A]
If I make a .xlam addin then I can call from `foo.xlsm` the functions defined in that addin `bar.xlam` as long as there is no `Option Private Module`. Is it possible to call functions in standard modules of an activex dll, or do we need a 3rd build type "COM DLL" for that? I.e. ActiveX is just defining objects you can `New` up?

I don't know much about this typelib stuff but it looks like ODL which is the language used to define typelibs has a [`module`](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/automat/module) definition type which my twinbasic activex package doesn't use (in olewoo.exe)

**Comment by bclothier** [comment]
If you decorate a function with `DllExport` attribute, then you can use the DLL without referencing it. 

[Here's an example of defining a exported DLL function returning an `Object`, allowing for referencing a COM object without actually referencing it via VBA's references](https://github.com/bclothier/TwinBasicSevenZip/blob/main/TwinProject/Sources/Exports.twin#L40-L43).

**Reply by Greedquest** [reply]
Ah thankyou, although that's what I've [already been using](https://github.com/Greedquest/vbInvoke/blob/6a9cf75f6cf2fd2ed9a4812a51939c10d38a5d17/twinproj/Sources/testing/ProxyObject.twin#L50-L55) for development. But that only works for "Standard DLL" build type correct? I'm after the ActiveX build type because then I get intellisense. I could use a predeclared class + methods but I think it should be possible to include standalone methods and modules in the typelib.

e.g. VBE7.dll has standalone modules and methods like `Math.Sin()`. Can I make 

```vba
Public Module Math
    Public Function Sin(ByVal x As Double) As Double
```
in a twinBASIC dll referenced from VBA, I'm not sure?

**Reply by bclothier** [reply]
Nope. The example above is an [ActiveX DLL](https://github.com/bclothier/TwinBasicSevenZip/blob/main/TwinProject/Settings#L4) so it does work whether it's standard or activeX DLL. 

The only caveat is that if you don't want to register, then you must work late-bound exclusively.

EDIT:

Ok, I think we're talking about two different things. If you want the intellisense and all that, then you need to register the library for VBIDE to know how to use it. You could then use the above trick to allow you to ship the DLL without requiring registration of that DLL but as said, it must work late-bound. Hope that helps?

**Reply by bclothier** [reply]
Late addendum: I forgot to point out that the modules aren't exported the same way. You can simulate one using a predeclared class, however.

**Reply by WaynePhillipsEA** [reply]
In type libraries, procedures you see in modules are just DLL declarations, exactly like writing your own `Declare` statements in VBx.  We don't yet offer any way to generate these Dll entries in type libraries.

A more COM friendly approach might be to use an `[AppObject]` class, which is basically like a predeclared class where you don't need to qualify its methods... So you can call just `DoSomething` on class `Foo` without the `Foo.` prefix.

**Reply by Greedquest** [reply]
Thanks both, DllExport + Declare I didn't realise was possible for ActiveX so that'll really help my development as I don't have to switch build types when debugging. @WaynePhillipsEA  AppObject looks good! Only downside is it pollutes my typelib and project with an extra "Application/Globals" class, and you can't have 1 appobject for each standard module, just a single appobject per dll.

Do you know what VBA does for the type library of a `.xlam`? Does it use the MIDL module type + convert everything to dll declarations? Or do they use appobjects? I ask because you can call standard module functions late-bound so I imagine modules are more like `IDispatch` objects and less like dll declares...

Should we open an issue for tracking `[DllExport]` in type libraries?
